server/actions_versions.go

```diff
@@ -156,6 +156,15 @@ func (s *RegistryServer) ListApiVersions(ctx context.Context, req *rpc.ListApiVe
 	if parent.ApiID != "-" {
 		q = q.Require("ApiID", parent.ApiID)
 	}
+	if parent.ProjectID != "-" && parent.ApiID != "-" {
+		if _, err := getApi(ctx, client, parent); err != nil {
+			return nil, err
+		}
+	} else if parent.ProjectID != "-" {
```

Comment:
I think some of my confusion reading this came from not realizing that parent.ApiID == "-" for this test (based on the condition above). It might help to restructure this so there's a single parent.ProjectID != "-" test wrapping the if condition on ApiId. 

Have you written this code for the more complicated case (specs)? It might be easier to work back from that.

Reply:
I think that works out fine for this case, but in specs I find it confusing with the different levels of nesting.
```
	if parent.ProjectID != "-" {
		if parent.ApiID != "-" {
			if parent.VersionID != "-" {
				if _, err := getVersion(ctx, client, parent); err != nil {
					return nil, err
				}
			} else {
				if _, err := getApi(ctx, client, parent.Api()); err != nil {
					return nil, err
				}
			}
		} else {
			if _, err := getProject(ctx, client, parent.Project()); err != nil {
				return nil, err
			}
		}
	}
```

I think my current approach with some additional details can increase clarity. Again with specs as an example.
```
	if parent.ProjectID != "-" && parent.ApiID != "-" && parent.VersionID != "-" {
		if _, err := getVersion(ctx, client, parent); err != nil {
			return nil, err
		}
	} else if parent.ProjectID != "-" && parent.ApiID != "-" && parent.VersionID == "-" {
		if _, err := getApi(ctx, client, parent.Api()); err != nil {
			return nil, err
		}
	} else if parent.ProjectID != "-" && parent.ApiID == "-" && parent.VersionID == "-" {
		if _, err := getProject(ctx, client, parent.Project()); err != nil {
			return nil, err
		}
	}
```